### PR TITLE
[Merged by Bors] - Separate Extract from Sub App Schedule

### LINKED ISSUE #6990: render app: Added stages get ignored
## Bevy version

0.9.1

## What you did

The use case (same as https://github.com/bevyengine/bevy/issues/6989): I want to have a uniform which is created based on 2 separate sets of extracted data. I'm bound on one side by the Extract stage where the extraction takes place, and the Prepare stage, where the instantiated uniform is written to the buffer via UniformComponentPlugin.

There is no stage between them. Adding a stage results in the system not running:

```
use bevy::prelude::*;
use bevy::render::{RenderApp, RenderStage};

fn main() {
    let mut app = App::new();
    app.add_plugins(DefaultPlugins);
    let render_app = app.get_sub_app_mut(RenderApp).unwrap();
    render_app
        .add_stage_after(RenderStage::Extract, "combine", SystemStage::single(system));
    app.run();
}


fn system() {
    println!("I want to break free");
}
```

## What went wrong

Expected: the system runs.

Actual: nothing gets printed.

### PR DESCRIPTION
**Pull request by hymm**
# Objective

- This pulls out some of the changes to Plugin setup and sub apps from #6503 to make that PR easier to review.
- Separate the extract stage from running the sub app's schedule to allow for them to be run on separate threads in the future
- Fixes #6990

## Solution

- add a run method to `SubApp` that runs the schedule
- change the name of `sub_app_runner` to extract to make it clear that this function is only for extracting data between the main app and the sub app
- remove the extract stage from the sub app schedule so it can be run separately. This is done by adding a `setup` method to the `Plugin` trait that runs after all plugin build methods run. This is required to allow the extract stage to be removed from the schedule after all the plugins have added their systems to the stage. We will also need the setup method for pipelined rendering to setup the render thread. See https://github.com/bevyengine/bevy/blob/e3267965e15f14be18eec942dcaf16807144eb05/crates/bevy_render/src/pipelined_rendering.rs#L57-L98

## Changelog

- Separate SubApp Extract stage from running the sub app schedule.

### SubApp `runner` has conceptually been changed to an `extract` function.

The `runner` no longer is in charge of running the sub app schedule. It's only concern is now moving data between the main world and the sub app. The `sub_app.app.schedule` is now run for you after the provided function is called.

```rust
// before
fn main() {
    let sub_app = App::empty();
    sub_app.add_stage(MyStage, SystemStage::parallel());
    
    App::new().add_sub_app(MySubApp, sub_app, move |main_world, sub_app| {
        extract(app_world, render_app);
        render_app.app.schedule.run();
    });
}

// after
fn main() {
        let sub_app = App::empty();
    sub_app.add_stage(MyStage, SystemStage::parallel());
    
    App::new().add_sub_app(MySubApp, sub_app, move |main_world, sub_app| {
        extract(app_world, render_app);
        // schedule is automatically called for you after extract is run
    });
}
```

### CODE REVIEW

**Comment by hymm** [comment]
It felt like these changes had merit on their own separate from pipelined rendering, so they could be reviewed separately.

**Review by james7132 — APPROVED** [review]
LGTM. My only concern is how this will work when stageless lands, but something like a dedicated schedule in Schedules works just fine.

**Comment by hymm** [comment]
> something like a dedicated schedule in Schedules works just fine.

I think that's my current preference. A separate extract schedule should work.

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -90,7 +90,20 @@ impl Debug for App {
 /// Each `SubApp` has its own [`Schedule`] and [`World`], enabling a separation of concerns.
 struct SubApp {
     app: App,
-    runner: Box<dyn Fn(&mut World, &mut App)>,
+    extract: Box<dyn Fn(&mut World, &mut App)>,
+}
+
+impl SubApp {
+    /// run the `SubApp`'s schedule
```
Comment (JoJoJet) [inline crates/bevy_app/src/app.rs:97]:
```suggestion
    /// Runs the `SubApp`'s schedule.
```

File: crates/bevy_app/src/app.rs [hunk 2]
```diff
@@ -90,7 +90,20 @@ impl Debug for App {
...
+    pub fn run(&mut self) {
+        self.app.schedule.run(&mut self.app.world);
+        self.app.world.clear_trackers();
+    }
+
+    /// extract data from main world to sub app
```
Comment (JoJoJet) [inline crates/bevy_app/src/app.rs:103]:
```suggestion
    /// Extracts data from main world to this sub-app.
```

File: crates/bevy_app/src/app.rs [hunk 3]
```diff
@@ -90,7 +90,20 @@ impl Debug for App {
 /// Each `SubApp` has its own [`Schedule`] and [`World`], enabling a separation of concerns.
 struct SubApp {
     app: App,
-    runner: Box<dyn Fn(&mut World, &mut App)>,
+    extract: Box<dyn Fn(&mut World, &mut App)>,
```
Comment (JoJoJet) [inline crates/bevy_app/src/app.rs:93]:
```suggestion
    // Extracts data from a world into a sub-app.
    extract: Box<dyn Fn(&mut World, &mut App)>,
```

File: crates/bevy_ecs/src/schedule/mod.rs [hunk 4]
```diff
@@ -361,6 +361,17 @@ impl Schedule {
             .and_then(|stage| stage.downcast_mut::<T>())
     }
 
+    /// Remove a [`Stage`] from the schedule
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/schedule/mod.rs:364]:
```suggestion
    /// Removes a [`Stage`] from the schedule.
```

File: crates/bevy_render/src/lib.rs [hunk 5]
```diff
@@ -92,6 +92,10 @@ pub enum RenderStage {
     Cleanup,
 }
 
+/// Resource for holding the extract stage of the rendering schedule
```
Comment (JoJoJet) [inline crates/bevy_render/src/lib.rs:95]:
```suggestion
/// Resource for holding the extract stage of the rendering schedule.
```

**Review by JoJoJet — APPROVED** [review]

**Comment by recatek** [comment]
Correct me if I'm wrong, but the switch to having only an extract method removes the authority of the sub app function to decide whether or not to update the sub-app at all, right? Do sub apps have run criteria to fill this gap? If not, could they?

**Comment by hymm** [comment]
> Correct me if I'm wrong, but the switch to having only an extract method removes the authority of the sub app function to decide whether or not to update the sub-app at all, right? Do sub apps have run criteria to fill this gap? If not, could they?

The way I would do this is to conditionally run a schedule from inside an exclusive system. This requires a bit of boilerplate right now, but should be significantly easier once stageless is merged. 

I could add support for a `runner` function that would be default run the schedule, but be replaceable. Though the use case for this is if you can't do what you need with just world and need direct access to the sub app. I thought most use cases for this would be taken care of by the extract function, but if there are other use cases that aren't covered I wouldn't might adding this.

**Comment by recatek** [comment]
The lost use case here isn't major, but I think it could be useful. That is, if you only want the SubApp to run sometimes (e.g. on a fixed update), and only want to perform the extract if it has run. You can do that under this setup by adding a resource with a frame counter on the SubApp and skipping the extract if that counter hasn't incremented, so it isn't a big deal, but it might be a nice ergonomics add to add a run criteria to the SubApp that also determines whether extract is called.

**Review by alice-i-cecile — APPROVED** [review]
bors r+

**Pull request closed without merging**